prusti-interface/src/environment/loops.rs

```diff
@@ -10,7 +10,7 @@ use crate::environment::procedure::BasicBlockIndex;
 use rustc_middle::mir;
 use rustc_middle::mir::visit::Visitor;
 use rustc_data_structures::graph::dominators::Dominators;
-use std::collections::{HashMap, HashSet};
+use rustc_hash::{FxHashMap as HashMap, FxHashSet as HashSet};
```

Comment:
I'd prefer to use `FxHashMap` and `FxHashSet` in the code; the diff gets bigger but more readable IMO (e.g. it's not necessary to jump to the top of the file to check if `HashMap` is really `HashMap` or `FxHashMap`).

Is there any reason to use `rustc_hash::{FxHashMap, FxHashSet` rather than `use rustc_data_structures::fx::{FxHashMap, FxHashSet}`?

Reply:
Thanks for the review!

> I'd prefer to use FxHashMap and FxHashSet in the code; the diff gets bigger but the code would be more readable IMO (e.g. it's not necessary to jump to the top of the file to check if HashMap is really HashMap or FxHashMap).

I had also considered this, but initially decided against it for two reasons:

1. The bigger diff has the potential to cause major merge conflicts for other very large PRs that are currently in progress. E.g. #799, #805 and probably others, too. I was worried that it would not be great "git etiquette" and that it might frustrate others when they get such a major merge conflict :smile: 
2. The [`FxHashMap`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_data_structures/fx/type.FxHashMap.html) and [`FxHashSet`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_data_structures/fx/type.FxHashSet.html) are just type aliases to `std::collections::{HashMap, HashSet}`. The last type parameter is different, though. Fx uses `BuildHasherDefault<FxHasher>` and `std` uses `RandomState` as the default type parameter. This means the APIs are identical for all but [2 methods](https://doc.rust-lang.org/1.57.0/std/collections/hash_map/struct.HashMap.html#implementations): `HashMap::with_capacity` and `HashMap::new` are only defined for HashMaps that have the `RandomState` type parameter (which is the reason I had to replace `HashMap::new()` with `HashMap::default()` everywhere). Since the API surface belongs to one and the same `HashMap`/`HashSet`, the times a developer needs to double check at the top would be minimal.

However, if it's still better for the maintainability to replace all instances with `FxHashMap` / `FxHashSet` everywhere, I'd be happy to make to change. I just wanted to run my motivation by you first to be sure that making the change will have the desired outcome :)

> Is there any reason to use rustc_hash::{FxHashMap, FxHashSet rather than rustc_data_structures::fx::{FxHashMap, FxHashSet}?

Good catch, there is no reason, I only just realized that the `rustc_data_structures` crate is provided by the compiler (it also depends on the `rustc_hash` crate, which is also provided by the compiler) when the `rustc_private` feature flag is enabled. I'll update this! Would it be acceptable to add `#![feature(rustc_private)]` to the `vir` and `viper` creates, too, or should we link it through `Cargo.toml` instead? It [looks like](https://github.com/rust-lang/rust/issues/27812) that feature flag won't be stabilized anytime soon, so it might be better to avoid it where it's not needed for other reasons.

Thanks!

Reply:
I'm not the one working on the other PRs, but the conflicts seem okay to me. Thanks for thinking at it :) Changing `HashMap::new` to `FxHashMap::default` vs `HashMap::default` should roughly be the same. There will be additional type declarations to be fixed, but they shouldn't be too many.

The API is the same, yes, but when looking at a PR it's easier to spot `HashMap` and comment "consider using FxHashMap because it's faster and more deterministic" rather than checking the `use ...` that might be or not in the diff. I tried to quantify this practice: [`rust-lang/rust`](https://grep.app/search?q=use%20.%2AFxHash.%2A%20as&regexp=true&filter[lang][0]=Rust&filter[repo][0]=rust-lang/rust), [`rust-lang/rust-analyzer`](https://grep.app/search?q=use%20.%2AFxHash.%2A%20as&regexp=true&filter[lang][0]=Rust&filter[repo][0]=rust-lang/rust-analyzer) and [`rust-lang/clippy`](https://grep.app/search?q=use%20.%2AFxHash.%2A%20as&regexp=true&filter[lang][0]=Rust&filter[repo][0]=rust-lang/clippy) never import `FxHashMap/Set as ...`, while [other smaller crates](https://grep.app/search?q=use%20.%2AFxHash.%2A%20as&regexp=true&filter[lang][0]=Rust) do. I would stick with the convention that the compiler developers use.

Reply:
Great, that works for me. I'll make the changes. Thanks! :)